internal/subscriber/core.go

```diff
@@ -39,36 +39,28 @@ func (c *Core) NewSubscriber(ctx context.Context,
 	ackCh chan *AckMessage,
 	modAckCh chan *ModAckMessage) (ISubscriber, error) {
 
-	topic := messagebroker.NormalizeTopicName(subscription.Topic)
-	retryTopic := messagebroker.NormalizeTopicName(subscription.GetRetryTopic())
-	dlqTopic := messagebroker.NormalizeTopicName(subscription.GetDeadLetterTopic())
-
 	groupID := subscription.Name
 
-	consumer, err := c.bs.GetConsumer(ctx, messagebroker.ConsumerClientOptions{Topics: []string{topic, retryTopic}, GroupID: groupID, GroupInstanceID: subscriberID})
+	consumer, err := c.bs.GetConsumer(ctx, messagebroker.ConsumerClientOptions{Topics: []string{subscription.Topic, subscription.GetRetryTopic()}, GroupID: groupID, GroupInstanceID: subscriberID})
 	if err != nil {
 		logger.Ctx(ctx).Errorw("subscriber: failed to create consumer", "error", err.Error())
 		return nil, err
 	}
 
-	retryProducer, err := c.bs.GetProducer(ctx, messagebroker.ProducerClientOptions{Topic: retryTopic, TimeoutMs: 500})
-	if err != nil {
-		logger.Ctx(ctx).Errorw("subscriber: failed to create retry producer", "error", err.Error())
-		return nil, err
-	}
-
-	dlqProducer, err := c.bs.GetProducer(ctx, messagebroker.ProducerClientOptions{Topic: dlqTopic, TimeoutMs: 500})
-	if err != nil {
-		logger.Ctx(ctx).Errorw("subscriber: failed to create dlq producer", "error", err.Error())
-		return nil, err
+	subsCtx, cancelFunc := context.WithCancel(ctx)
+	// using the subscriber ctx for retrier as well. This way when the ctx() for subscribers is done,
+	// all the delay-consumers spawned within retrier would also get marked as done.
+	var retrier IRetrier
+	if subscription.DelayConfig != nil {
+		retrier, err = NewRetrier(subsCtx, subscription.DelayConfig, c.bs, NewPushToPrimaryRetryTopicHandler(c.bs))
```

Comment:
do we need a retry per subscriber, can't we just have a retry module as singleton which manages all subscriptions and can retry a message based on message and subscription attributes

Reply:
We had decided during our Spec discussion that each subscriber/subscription should get its own retry module.
- This way the retry module handles only the subscription-specific traffic and one topic experiencing high retry messages will not affect the throughput of another subscription. 
- Additionally makes it easier to break for ingestion topics in the future.